### ISSUE TITLE
Changing from Utterance to Database View should keep the current frame in view

### ISSUE THREAD

**Issue by cneidle** [DCS-LCSR/SignStream3#631]
But that's not what happens.  When you go from Utterance to Database view, what shoes in the Database view is whatever had been previously viewed in the Database view.  For example, I was editing SALUTE.  Then I go to the utterance view.  I switch to a different utterance (BLOW), and then I go back to Database view, and I am brought back to SALUTE.  (These are arranged alphabetically, so these are very far apart.)

<img width="363" alt="Screen Shot 2023-02-02 at 10 42 19 PM" src="https://user-images.githubusercontent.com/13629122/216507610-9084ccbc-986a-41c7-98af-114255256ba2.png">

<img width="329" alt="Screen Shot 2023-02-02 at 10 42 27 PM" src="https://user-images.githubusercontent.com/13629122/216507612-1b9529ed-fe23-4b8d-96bb-e8b8bb474cc3.png">

The green alignment indicator is where it should be:  at BLOW.  But that is not in view (although I can see where the green alignment indicator is by zooming out):

<img width="581" alt="Screen Shot 2023-02-02 at 10 45 29 PM" src="https://user-images.githubusercontent.com/13629122/216507840-a1c8da84-3658-4d3b-8af7-9245ad6601d3.png">

**Comment by gregorydimitriadis** [comment]
This is from the "Go to Utterance View" and "Back to Database View"  buttons, I'm assuming? I believe (not sure since I don't think I did it) that those were not coded to operate on the green indicator line, but to remember the previous state of being in those views and go to them.

**Comment by cneidle** [comment]
The green line remains where it was in the utterance window, as it should, when the database window opens, but the green line (i.e., current frame) is not in view.
The current frame should always be in view.
This behavior did not occur in prior versions of SignStream.

**Comment by cneidle** [comment]
Keeping the video alignment indicator in the field of view does not seem like an enhancement to me... FWIW.

**Comment by douglas-motto-at-rutgers** [comment]
"The current frame should always be in view."

If meaning...the green bar/frame should always be in the viewable window...I do not believe this is always true.  You can have the green bar in the viewable window and then ("not" using the arrow keys) scroll or slide the viewable window to the right or left with the mouse.  The green bar/frame remains where is was and the viewable window may no long include the green bar/frame in that time period.

The green bar is always in sync with the video frame.  The green bar may move by arrow keys, changing the slider in the video player (playing or going to beginning or end of video), or selecting an Utterance in the utterance index.

If the green bar is moved, the viewable window moves to where the green bar is.

But not the other way around.

**Comment by douglas-motto-at-rutgers** [comment]
The green bar is also moved when entering the "Utterance" view from the "Database" view by double clicking on an utterance.

**Comment by douglas-motto-at-rutgers** [comment]
In terms of what the "Go To Utterance" or "Go To Database" view does...

Clicking that button returns the the previous viewable window the "Utterance" or "Database" view was last at.  It's intended as a "return as previously seen".

Clicking "Go to Utterance" before the "Utterance" view has been visited, will have a window shown including the entire video.

Double clicking on an utterance in the "Database" view will open the "Utterance" view showing that utterance.

When "Unlocked" the zoom level will change automatically when clicking on an utterance from the database view, or clicking on an utterance in the utterance index of the "Utterance" view...in order to display that entire utterance.

When the zoom level is "Locked" the zoom level is remembered and not changed automatically.  The Utterance view and the Database view have their own independently different "Locked" zoom levels.

**Comment by douglas-motto-at-rutgers** [comment]
So if you...

1)Unlock the zoom.
2)Double click on an utterance from the "Utterance" view.
3)You should see that entire utterance.
4)Move the green bar to the middle of that utterance.
5)Click "Go to Database" view.
6)Move the green bar outside of the utterance.
7)Click "Go to Utterance" view.

You will no longer see the green bar in the viewable window because that window is the same as seen in #3...that last window visited in the "Utterance" view.

**Comment by douglas-motto-at-rutgers** [comment]
Note: I use the term "window" to present the time window...or the range of viewable time shown from start to end on the screen.

**Comment by douglas-motto-at-rutgers** [comment]
There was one possible issue with the zoom Lock and Unlock.  That toggle button to lock/unlock is shared between the "Database" and "Utterance" views.

I believe it was originally intended to be independently toggable...Ex: Locked for the "Database" view, but Unlocked for the "Utterance" view...and vice versa.

**Comment by douglas-motto-at-rutgers** [comment]
At one point these settings (viewed time windows, zoom levels, and zoom un/locked) were being worked on to be remembered between saving, closing and reopening a collection.  Saving these settings within the collection file themselves.

But I don't know if this was completed.

**Comment by cneidle** [comment]
Thanks - but see below.

> On Feb 10, 2023, at 5:34 PM, douglas-motto-at-rutgers ***@***.***> wrote:
> 
> 
> In terms of what the "Go To Utterance" or "Go To Database" view does...
> 
> Clicking that button returns the the previous viewable window the "Utterance" or "Database" view was last at. It's intended as a "return as previously seen"
> 

That is absolutely not the desired function.  I don’t know when/how that was decided for Go to Database.

Go to Database should take you to the same frame in the database view as what was being viewed in the Utterance window.

There is no plausible use case where a user would want to go from a frame in an Utterance to some previous position in the Database view.

> Clicking "Go to Utterance" before the "Utterance" view has been visited, will have a window shown including the entire video.
> 
> Double clicking on an utterance in the "Database" view will open the "Utterance" view showing that utterance.
> 
> 

Yes - It should open that utterance, at the start of the utterance.

**Comment by cneidle** [comment]
If you go to Database view, the green alignment indicator, which is where it is
supposed to be, should be in view.

The user can move it, but this is what the default view should be.

**Comment by cneidle** [comment]
This is really, really annoying.  It would be excellent if we could get this fixed.  When you go from Utterance view to Database view, the current frame should be retained and visible.

**Comment by gregorydimitriadis** [comment]
To note: this could have something to do with the zoom value when the 'return' action occurs. Whether the current frame remains in the visible portion or not.

**Comment by gregorydimitriadis** [comment]
@cneidle do we wish to have the same functionality when we do a 'return to utterance view'? Specifically that button from the top-left of the screen, but not the action of double-clicking an individual utterance - which would retain the current functionality of jumping to the start of that utterance.

**Comment by cneidle** [comment]
Ok to leave the functionality of that button as is.  Thanks!

**Comment by cneidle** [comment]
This basically seems to work now.  Thank you!  I know we discussed that the "Go to utterance button" could be left with the most recent functionality, going **back** to the previous location in the utterance window.  However, there is something very odd about having the button in the utterance window as "**Back** to Database View" and the one in the database view saying "**Go** to Utterance View".  Apologies for changing my mind, but upon using it, it is clear to me that it would be MUCH better if the "Go to Utterance View" would go to the current location from the database view.  If that is too hard to change, then the button should be relabeled as "Go (back) to Utterance View."

**Comment by gregorydimitriadis** [comment]
So the change should actually already include what is requested here - that the "back to utterance view" snaps to the current location of the green slider. Is that observable here?

Also - was this intentionally typed or a typo? This seems odd:

"RIT_P02-s01.ssc to Utterance View"

**Comment by cneidle** [comment]
Yes  - the "back to database view" should take the user to the current frame but in database view, and that frame should be in the viewable range.  I believe that is happening now, correctly.

However there is something very weird about having the button in the database view "Go to Utterance view" *not* do the same thing, i.e., *not* take the user to the current frame in the utterance window.  Having it remember where the user last was in utterance view is unexpected behavior.  If that behavior is to remain, then we should at least change the label to "go (back) to utterance view" -- but it would be better to simply have it take the user to the current frame in the utterance view.

**Comment by cneidle** [comment]
I corrected the typo you pointed out by editing the original comment.  Sorry about that.

**Comment by cneidle** [comment]
This seems to work now :-)  Thanks
